portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -117,15 +138,50 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     // Get the blob key from Blobstore (if blob was submitted).
     // "image" is the name of the file input in the comment form.
-    String imageKey = getImageKey(request, "image");
+    BlobKey imageKey = getImageKey(request, "image");
+   
+    // Use one annotate image request to get the labels associated with and the objects
+    // detected in the image uploaded by the user, if there was one.
+    // Also, create a string from the image key if there was an image.
+    ArrayList<String> labels = new ArrayList<String>();
+    ArrayList<Float> labelScores = new ArrayList<Float>();
+    ArrayList<String> objects = new ArrayList<String>();
+    ArrayList<Float> objScores = new ArrayList<Float>();
+    String imKey = "";
+
+    if (imageKey != null) { 
+      // Create string from key to store in datastore.
+      imKey = imageKey.getKeyString(); 
+
+      byte[] imageBytes = getBlobBytes(imageKey);
+      List<LocalizedObjectAnnotation> imageObjects = getImageObjects(imageBytes);
+      List<EntityAnnotation> imageLabels = getImageLabels(imageBytes);
+      
+      // Build the lists of image descriptions and scores.
+      for (EntityAnnotation label: imageLabels) {
```

Comment:
Both imageLabels and imageObjects can be `null` - this will cause an NullPointerException.